src/editor/codelens.ts

```diff
@@ -0,0 +1,115 @@
+import {
+    CancellationToken,
+    CodeLens,
+    CodeLensProvider,
+    Event,
+    EventEmitter,
+    ExtensionContext,
+    Range,
+    TextDocument,
+    languages,
+    workspace
+} from 'vscode';
+import { ExtensionApi } from '../backend';
+import { AnalysisPathEvent, AnalysisPathKind } from '../backend/types';
+
+export class CodeLensStepsProvider implements CodeLensProvider {
+    constructor(ctx: ExtensionContext) {
+        ctx.subscriptions.push(this._onDidChangeCodeLenses = new EventEmitter());
+
+        ExtensionApi.diagnostics.diagnosticsUpdated(
+            this._onDidChangeCodeLenses.fire,
+            this._onDidChangeCodeLenses,
+            ctx.subscriptions
+        );
+
+        ctx.subscriptions.push(languages.registerCodeLensProvider('*', this));
+    }
+
+    private _onDidChangeCodeLenses: EventEmitter<void>;
+    public get onDidChangeCodeLenses(): Event<void> {
+        return this._onDidChangeCodeLenses.event;
+    }
+
+    provideCodeLenses(document: TextDocument, _token: CancellationToken): CodeLens[] {
+        if (ExtensionApi.diagnostics.selectedEntry === undefined ||
+            !workspace.getConfiguration('codechecker.editor').get('enableCodeLens')) {
+            return [];
+        }
+
+        const entry = ExtensionApi.diagnostics.selectedEntry;
+        const fullPath = entry.diagnostic.path
+            .filter(elem => elem.kind === AnalysisPathKind.Event) as AnalysisPathEvent[];
+
+        if (fullPath.length === 0) {
+            return [];
+        }
+
+        const codeLenses = [];
+
+        // Used when consecutive steps are displayed
+        let previousJumpIdx = -1;
+
+        for (const [idx, pathItem] of fullPath.entries()) {
+            if (entry.diagnostic.files[pathItem.location.file] !== document.uri.fsPath) {
+                continue;
+            }
+
+            // TODO: Handle multiple ranges
+            const range = new Range(
+                pathItem.location.line-1,
+                0,
+                pathItem.location.line-1,
+                0,
+            );
+
+            // FIXME: Handle multiple steps in the same line
+            codeLenses.push(new CodeLens(range, {
+                'title': `${idx + 1}: ${pathItem.message}`,
```

Comment:
In the tree view for a division by zero error I see that I have 3 bug steps:
![image](https://user-images.githubusercontent.com/6695818/143427203-f0a90974-c292-4be0-848d-7fa4650b295a.png)

But on the source code CodeLens shows me an extra bug step:
![image](https://user-images.githubusercontent.com/6695818/143427313-3522a0cf-8845-4930-b662-6cc1d62e13af.png)

It would be good to show this extra bug step in the tree view too (we are doing the same thing in the CodeChecker GUI). We can create a separate issue for it if we don't want to fix this in this PR.

Reply:
Most likely a leftover from #26, should be fixed by the rebase.
Let me know if you still encounter it.

Reply:
I still have this issue. My example project is the following:
**main.h**:
```h
#ifndef MAIN_H
#define MAIN_H

double foo(int param) {
  return 1 / param;
}

#endif
```

**main.cpp**:
```cpp
#include "main.h"

int main() {
  return foo(0);
}
```

![image](https://user-images.githubusercontent.com/6695818/144193280-cf2c23de-b798-4bec-887b-0d2d74cd9379.png)
![image](https://user-images.githubusercontent.com/6695818/144193309-9eeba43f-9e72-4249-a282-4b8e71c1e68c.png)

Reply:
I created a separate issue for this problem: #36.